### PR TITLE
Improve interrupt handling

### LINKED ISSUE #127: What's the current status of IRQs?
There is currently no good story for how IRQs work, and the code for the assembly part of IRQs handlers is commented out. Is this a mistake, or are there further plans for how to handle this in the future?

There used to be a very complex wrapper around IRQs, and I had assumed that most of it was needed for one reason or another. Does it turn out not to be?

### LINKED ISSUE #152: Many vblank irqs sent per frame
I'm not sure if this is an error in the library or my code, but I'm thoroughly confused by this. 
Consider the following code:
```rust
#![no_std]
#![no_main]
#![feature(isa_attribute)]

use gba::{fatal, prelude::*};

#[panic_handler]
fn panic(info: &core::panic::PanicInfo) -> ! {
    fatal!("{}", info);
}

fn wait_vblank() {
    unsafe { gba::bios::VBlankIntrWait() }
}

fn init_registers() {
    unsafe {
        IME.write(true);
        IE.write(InterruptFlags::new().with_vblank(true));
    }

    DISPSTAT.write(DisplayStatus::new().with_vblank_irq_enabled(true));
}

#[instruction_set(arm::a32)]
extern "C" fn irq_handler_a32() {
    static COUNTER: Static<usize> = Static::new(0);
    let c = COUNTER.read();
    gba::warning!("{}", c);
    COUNTER.write(c + 1);
    unsafe { INTR_WAIT_ACKNOWLEDGE.write(INTR_WAIT_ACKNOWLEDGE.read().with_vblank(true)) }
}

#[no_mangle]
fn main() -> ! {
    unsafe { USER_IRQ_HANDLER.write(Some(irq_handler_a32)) };
    init_registers();
    loop {
        wait_vblank();
    }
}

```
All this should do is print the value of `COUNTER` once every frame, after the vblank irq is raised. Instead it seems to run the irq handler many times per frame (on my laptop around 5 times per frame in debug mode and 150 times per frame in release mode). I'm pretty stumped by this and can't think of an explanation as for why multiple vblank interrupts should be fired each frame. If I replace the `wait_vblank` implementation with one which just does a busy loop, everything works as intended, but I don't really want to do that.

### PR DESCRIPTION
**Pull request by Ruin0x11**
- Replaces the main interrupt handler with a new "switchboard" design that dispatches based on interrupt kind, mostly inspired by tonclib.
- Allow setting an individual interrupt handler function per interrupt kind.
- Add some convenience functions for enabling/disabling the triggering of each interrupt.
- Add documentation about the interrupt handling features.

Closes #152. Closes #127.

### CODE REVIEW

File: src/rsrt0.S [hunk 1]
```diff
@@ -48,57 +48,88 @@ __start:
...
+    bic r2, r2, #0xDF
+    orr r2, r2, #0x1F
+    msr cpsr, r2
+
+    @ Enable interrupts and jump to handler (Thumb)
+    stmfd sp!, {r3, lr}   @ sp_sys, {&REG_IE, lr_sys}
+    mov r2, #1
+    str r2, [r3, #8]    @ REG_IME=1
+    adr lr, .Lpost_isr
+    bx  r0
```
Comment (Lokathor) [inline src/rsrt0.S:114]: it looks like you're jumping to a handler using `bx`, so why do the handler pointers have to be t32 code? Doesn't the `bx` allow jumping from `a32` to other `a32` code as well?

File: src/irq.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    IrqSender { offset: 0x0000, mask: 0x0000 }, // Gamepak (none)
+];
+
+const MMIO_BASE_ADDRESS: usize = 0x0400_0000;
+
+#[inline]
+unsafe fn set_irq_flag(kind: IrqKind) {
+    let sender = &SENDERS[kind as usize];
+    let addr: *mut u16 = core::mem::transmute(MMIO_BASE_ADDRESS + sender.offset);
+    *addr |= sender.mask;
```
Comment (Lokathor) [inline src/irq.rs:114]: this needs to use volatile accesses.

File: src/irq.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    IE.write(InterruptFlags(IE.read().0 | (1 << (kind as u8))))
+}
+
+#[inline]
+unsafe fn unset_irq_flag(kind: IrqKind) {
+    let sender = &SENDERS[kind as usize];
+    let addr: *mut u16 = core::mem::transmute(MMIO_BASE_ADDRESS + sender.offset);
+    *addr &= !sender.mask;
```
Comment (Lokathor) [inline src/irq.rs:122]: this also needs to use volatile accesses.

**Review by Lokathor — CHANGES_REQUESTED** [review]

File: src/rsrt0.S (hunk 1 above)
Comment (Ruin0x11) [reply src/rsrt0.S:115]: The mode can be changed depending on the least significant bit of the address, if it's 1 then Thumb mode is used. I wasn't setting it actually so I fixed that.

Comment (Lokathor) [reply src/rsrt0.S:115]: I think you misunderstand a bit: all thumb functions will already have an odd address assigned to their symbol by the linker. Just using `bx r0` with no address manipulation will already handle any ARM/Thumb exchange that needs to happen.

**Comment by Hpmason** [comment]
While we're adding better IRQ handling, would it be a good idea to work on #150? The examples seem to be the only problem in the CI. I was able to build the examples by adding the new IRQ handling to the examples.

**Comment by Lokathor** [comment]
sounds reasonable.

**Comment by Lokathor** [comment]
This is now very stale, I'll close it.

**Pull request closed without merging**